web-api/src/storage/elastic.rs

```diff
@@ -322,3 +326,186 @@ struct IngestedDocument<'a> {
     embedding: &'a NormalizedEmbedding,
     tags: &'a [DocumentTag],
 }
+
+struct KnnSearchParts {
+    knn_object: JsonObject,
+    generic_parameters: JsonObject,
+    inner_filter: JsonObject,
+}
+
+impl KnnSearchParams<'_> {
+    fn create_common_knn_search_parts(&self) -> KnnSearchParts {
+        let inner_filter = self.create_search_filter();
+        let knn_object = self.create_knn_request_object(&inner_filter);
+
+        let mut generic_parameters = json_object!({
+            "size": self.count
+        });
+
+        if let Some(min_score) = self.min_similarity {
+            generic_parameters.extend(json_object!({
+                "min_score": min_score,
+            }));
+        }
+
+        KnnSearchParts {
+            knn_object,
+            generic_parameters,
+            inner_filter,
+        }
+    }
+
+    fn create_search_filter(&self) -> JsonObject {
+        let mut filter = JsonObject::new();
+        if self.excluded.is_empty().not() {
```

Comment:
❓ this is quite odd, is it on purpose? otherwise we can write it in the usual way with the operator instead and also remove the `Not` import.
```suggestion
        if !self.excluded.is_empty() {
```